File: morphir/src/org/finos/morphir/ir/gen1/TypeSpecification.scala
package org.finos.morphir.ir.gen1

import org.finos.morphir.ir.gen1.naming.*

sealed trait TypeSpecification[+A] { self =>
  import TypeSpecification.*

  final def properties: Properties[A] =
    self match {
      case TypeAliasSpecification(_, expr)              => Properties.TypeAlias(expr)
      case CustomTypeSpecification(_, ctors)            => Properties.CustomType(ctors)
      case DerivedTypeSpecification(_, derivationProps) => derivationProps
      case OpaqueTypeSpecification(_)                   => Properties.OpaqueType
    }

  final def map[B](f: A => B): TypeSpecification[B] =
    self match {
      case spec @ TypeAliasSpecification(_, _)  => TypeAliasSpecification(spec.typeParams, spec.expr.map(f))
      case spec @ OpaqueTypeSpecification(_)    => spec
      case spec @ CustomTypeSpecification(_, _) => spec.copy(ctors = spec.ctors.map(f))
      case spec @ DerivedTypeSpecification(_, _) =>
        val baseType = spec.derivationProps.baseType.map(f)
        val props    = spec.derivationProps.copy(baseType = baseType)
        spec.copy(derivationProps = props)
    }

  def typeParams: List[Name]

}

object TypeSpecification {
  final case class TypeAliasSpecification[+A](typeParams: List[Name], expr: Type[A]) extends TypeSpecification[A]
  final case class OpaqueTypeSpecification(typeParams: List[Name])                   extends TypeSpecification[Nothing]
  final case class CustomTypeSpecification[+A](typeParams: List[Name], ctors: TypeConstructors[A])
      extends TypeSpecification[A]

  final case class DerivedTypeSpecification[+A](typeParams: List[Name], derivationProps: Properties.DerivedType[A])
      extends TypeSpecification[A]

  sealed trait Properties[+A] { self =>
    def map[B](f: A => B): Properties[B] =
      self match {
        case props @ Properties.TypeAlias(_)         => Properties.TypeAlias(props.expr.map(f))
        case props @ Properties.CustomType(_)        => props.copy(ctors = props.ctors.map(f))
        case props @ Properties.DerivedType(_, _, _) => props.copy(baseType = props.baseType.map(f))
        case _                                       => Properties.OpaqueType
      }
  }

  object Properties {
    final case class TypeAlias[+A](expr: Type[A]) extends Properties[A]
    type OpaqueType = Properties.OpaqueType.type
    case object OpaqueType extends Properties[Nothing] {
      type Type[+A] = Properties.OpaqueType.type
    }
    final case class CustomType[+A](ctors: TypeConstructors[A]) extends Properties[A]
    final case class DerivedType[+A](baseType: Type[A], fromBaseType: FQName, toBaseType: FQName)
        extends Properties[A]
  }
}

File: pkg/models/ir/type_specification.go
package ir

// TypeSpecification corresponds to Morphir.IR.Type.Specification.
//
// The name is prefixed to avoid collisions with other IR modules (e.g. ModuleSpecification).
type TypeSpecification[A any] interface {
	isTypeSpecification()
	TypeParams() []Name
}

// TypeAliasSpecification corresponds to: TypeAliasSpecification (List Name) (Type a)
type TypeAliasSpecification[A any] struct {
	params []Name
	exp    Type[A]
}

func NewTypeAliasSpecification[A any](params []Name, exp Type[A]) TypeSpecification[A] {
	var copied []Name
	if len(params) > 0 {
		copied = make([]Name, len(params))
		copy(copied, params)
	}
	return TypeAliasSpecification[A]{params: copied, exp: exp}
}

func (TypeAliasSpecification[A]) isTypeSpecification() {
	// Marker method for the TypeSpecification sum type.
}

func (s TypeAliasSpecification[A]) TypeParams() []Name {
	if len(s.params) == 0 {
		return nil
	}
	copied := make([]Name, len(s.params))
	copy(copied, s.params)
	return copied
}

func (s TypeAliasSpecification[A]) Expression() Type[A] {
	return s.exp
}

// OpaqueTypeSpecification corresponds to: OpaqueTypeSpecification (List Name)
type OpaqueTypeSpecification[A any] struct {
	params []Name
}

func NewOpaqueTypeSpecification[A any](params []Name) TypeSpecification[A] {
	var copied []Name
	if len(params) > 0 {
		copied = make([]Name, len(params))
		copy(copied, params)
	}
	return OpaqueTypeSpecification[A]{params: copied}
}

func (OpaqueTypeSpecification[A]) isTypeSpecification() {
	// Marker method for the TypeSpecification sum type.
}

func (s OpaqueTypeSpecification[A]) TypeParams() []Name {
	if len(s.params) == 0 {
		return nil
	}
	copied := make([]Name, len(s.params))
	copy(copied, s.params)
	return copied
}

// CustomTypeSpecification corresponds to: CustomTypeSpecification (List Name) (Constructors a)
type CustomTypeSpecification[A any] struct {
	params       []Name
	constructors TypeConstructors[A]
}

func NewCustomTypeSpecification[A any](params []Name, ctors TypeConstructors[A]) TypeSpecification[A] {
	var copiedParams []Name
	if len(params) > 0 {
		copiedParams = make([]Name, len(params))
		copy(copiedParams, params)
	}
	var copiedCtors TypeConstructors[A]
	if len(ctors) > 0 {
		copiedCtors = make(TypeConstructors[A], len(ctors))
		copy(copiedCtors, ctors)
	}
	return CustomTypeSpecification[A]{params: copiedParams, constructors: copiedCtors}
}

func (CustomTypeSpecification[A]) isTypeSpecification() {
	// Marker method for the TypeSpecification sum type.
}

func (s CustomTypeSpecification[A]) TypeParams() []Name {
	if len(s.params) == 0 {
		return nil
	}
	copied := make([]Name, len(s.params))
	copy(copied, s.params)
	return copied
}

func (s CustomTypeSpecification[A]) Constructors() TypeConstructors[A] {
	if len(s.constructors) == 0 {
		return nil
	}
	copied := make(TypeConstructors[A], len(s.constructors))
	copy(copied, s.constructors)
	return copied
}

// DerivedTypeSpecificationDetails corresponds to:
// { baseType : Type a, fromBaseType : FQName, toBaseType : FQName }
type DerivedTypeSpecificationDetails[A any] struct {
	baseType     Type[A]
	fromBaseType FQName
	toBaseType   FQName
}

func DerivedTypeSpecificationDetailsFromParts[A any](baseType Type[A], fromBaseType FQName, toBaseType FQName) DerivedTypeSpecificationDetails[A] {
	return DerivedTypeSpecificationDetails[A]{
		baseType:     baseType,
		fromBaseType: fromBaseType,
		toBaseType:   toBaseType,
	}
}

func (d DerivedTypeSpecificationDetails[A]) BaseType() Type[A] {
	return d.baseType
}

func (d DerivedTypeSpecificationDetails[A]) FromBaseType() FQName {
	return d.fromBaseType
}

func (d DerivedTypeSpecificationDetails[A]) ToBaseType() FQName {
	return d.toBaseType
}

// DerivedTypeSpecification corresponds to: DerivedTypeSpecification (List Name) DerivedTypeSpecificationDetails
type DerivedTypeSpecification[A any] struct {
	params  []Name
	details DerivedTypeSpecificationDetails[A]
}

func NewDerivedTypeSpecification[A any](params []Name, details DerivedTypeSpecificationDetails[A]) TypeSpecification[A] {
	var copied []Name
	if len(params) > 0 {
		copied = make([]Name, len(params))
		copy(copied, params)
	}
	return DerivedTypeSpecification[A]{params: copied, details: details}
}

func (DerivedTypeSpecification[A]) isTypeSpecification() {
	// Marker method for the TypeSpecification sum type.
}

func (s DerivedTypeSpecification[A]) TypeParams() []Name {
	if len(s.params) == 0 {
		return nil
	}
	copied := make([]Name, len(s.params))
	copy(copied, s.params)
	return copied
}

func (s DerivedTypeSpecification[A]) Details() DerivedTypeSpecificationDetails[A] {
	return s.details
}
